tiledb/api/src/query/read/output/arrow.rs

```diff
@@ -0,0 +1,340 @@
+    for PrimitiveArray<<C as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>
+where
+    C: ArrowPrimitiveTypeNative,
+    PrimitiveArray<<C as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>:
+        From<Vec<C>> + From<Vec<Option<C>>>,
+{
+    type Error = crate::error::Error;
+
+    fn try_from(value: RawReadOutput<C>) -> TileDBResult<Self> {
+        type MyPrimitiveArray<C> =
+            PrimitiveArray<<C as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>;
+
+        if value.input.cell_offsets.is_some() {
+            return Err(Error::Datatype(DatatypeErrorKind::ExpectedFixedSize(
+                None,
+            )));
+        }
+
+        Ok(if let Some(validity) = value.input.validity {
+            let data: Vec<C> = match value.input.data {
+                Buffer::Empty => vec![],
+                Buffer::Owned(data) => {
+                    let mut data = data.into_vec();
+                    data.truncate(value.nvalues);
+                    data
+                }
+                Buffer::Borrowed(data) => data[0..value.nvalues].to_vec(),
+            };
+
+            let validity = match validity {
+                Buffer::Empty => vec![],
+                Buffer::Owned(v) => {
+                    let mut v = v.into_vec();
+                    v.truncate(value.nvalues);
+                    v
+                }
+                Buffer::Borrowed(v) => v[0..value.nvalues].to_vec(),
+            };
+            let validity = validity
+                .into_iter()
+                .map(|v| v != 0)
+                .collect::<arrow::buffer::NullBuffer>();
+            MyPrimitiveArray::<C>::new(data.into(), Some(validity))
+        } else {
+            let mut v: Vec<C> = match value.input.data {
+                Buffer::Empty => vec![],
+                Buffer::Owned(b) => b.into_vec(),
+                Buffer::Borrowed(b) => b.to_vec(),
+            };
+            v.truncate(value.nvalues);
+            v.into()
+        })
+    }
+}
+
+impl<'data, C> From<RawReadOutput<'data, C>> for Arc<dyn ArrowArray>
+where
+    C: ArrowPrimitiveTypeNative,
+    PrimitiveArray<<C as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>:
+        From<Vec<C>> + From<Vec<Option<C>>>,
+{
+    fn from(mut value: RawReadOutput<'data, C>) -> Self {
+        /* TODO: needs the cell val num to make the decision properly */
+
+        type MyPrimitiveArray<C> =
+            PrimitiveArray<<C as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>;
+
+        let offsets = value.input.cell_offsets.take();
+
+        let mut v_offsets = if let Some(offsets) = offsets {
+            match offsets {
+                Buffer::Empty => vec![],
+                Buffer::Owned(offsets) => {
+                    let mut offsets = offsets.into_vec();
+                    offsets.truncate(value.nvalues);
+                    offsets
+                }
+                Buffer::Borrowed(offsets) => offsets[0..value.nvalues].to_vec(),
```

Comment:
I've got no idea how often this is called, but this strikes me as wrong by design. There shouldn't be allocations involved here. These can and will end up being gigabyte sized allocations when users eventually start banging on this. I don't think we need to change it now, but long term we'll definitely need to revisit this.

Reply:
If the user owns their query buffer, then this is unavoidable (without being unsafe).  Arrow either takes ownership of the buffer (as we see just above this), or copies.

If the user moves their buffer into the query, which we don't have an option for yet, or if they let tiledb manage the buffer, then we could end up with `Buffer::Owned` and avoid the copy.  I don't think that's actually implemented yet, but it is at least a possibility.